src/xdp/program.c

```diff
@@ -1387,6 +1562,112 @@ XdpProgramTraceObject(
     }
 }
 
+static const ebpf_context_descriptor_t EbpfXdpContextDescriptor = {
+    .size = sizeof(xdp_md_t),
+    .data = FIELD_OFFSET(xdp_md_t, data),
+    .end = FIELD_OFFSET(xdp_md_t, data_end),
+    .meta = FIELD_OFFSET(xdp_md_t, data_meta),
+};
+
+#define XDP_EXT_HELPER_FUNCTION_START EBPF_MAX_GENERAL_HELPER_FUNCTION
+
+// XDP helper function prototype descriptors.
+static const ebpf_helper_function_prototype_t EbpfXdpHelperFunctionPrototype[] = {
+    {
+        .helper_id = XDP_EXT_HELPER_FUNCTION_START + 1,
+        .name = "bpf_xdp_adjust_head",
+        .return_type = EBPF_RETURN_TYPE_INTEGER,
+        .arguments = {
+            EBPF_ARGUMENT_TYPE_PTR_TO_CTX,
+            EBPF_ARGUMENT_TYPE_ANYTHING,
+        },
+    },
+};
+
+// TODO: fix and upstream const-correct definitions
+#pragma warning(suppress:4090) // 'initializing': different 'const' qualifiers
+static const ebpf_program_info_t EbpfXdpProgramInfo = {
+// TODO: fix and upstream const-correct definitions
+#pragma warning(suppress:4090) // 'initializing': different 'const' qualifiers
+    .program_type_descriptor = {
+        .name = "xdp",
+        .context_descriptor = &EbpfXdpContextDescriptor,
+        .program_type = EBPF_PROGRAM_TYPE_XDP_INIT,
+        BPF_PROG_TYPE_XDP,
+    },
+    .count_of_program_type_specific_helpers = RTL_NUMBER_OF(EbpfXdpHelperFunctionPrototype),
+    .program_type_specific_helper_prototype = EbpfXdpHelperFunctionPrototype,
+};
+
+static
+int
+EbpfXdpAdjustHead(
+    _Inout_ xdp_md_t *Context,
+    _In_ int Delta
+    )
+{
+    //
+    // Not implemented.
+    //
+    UNREFERENCED_PARAMETER(Context);
+    UNREFERENCED_PARAMETER(Delta);
+    return -1;
```

Comment:
Does `-1` mean failure? Annotate as such?

Reply:
Maybe a "must check result" annotation too?

Reply:
I've added a comment, but since this is only invoked via a dispatch table defined by eBPF and its verifier, I don't see a ton of value in SAL annotation.